dokka-integration-tests/gradle/src/main/kotlin/org/jetbrains/dokka/it/gradle/junit/TestedVersions.kt

```diff
@@ -0,0 +1,228 @@
+
+    constructor(
+        kgp: String,
+        gradle: String,
+        dgp: String,
+        agp: String?,
+        composeGradlePlugin: String?,
+    ) : this(
+        kgp = SemVer(kgp),
+        gradle = SemVer(gradle),
+        dgp = SemVer(dgp),
+        agp = agp?.let { SemVer(it) },
+        composeGradlePlugin = composeGradlePlugin?.let { SemVer(it) },
+    )
+
+    /**
+     * Check if the combination of versions is valid.
+     *
+     * For example, AGP 7 must use Gradle 7. If [agp] is `8.0.0` and [gradle] is `7.0.0`
+     * then the combination is invalid.
+     */
+    private fun isValid(): Boolean {
+        if (agp != null) {
+            if (agp.major != gradle.major) {
+                return false
+            }
+
+            when (kgp) {
+                SemVer("1.9.0") -> {
+                    if (agp !in SemVer("7.0.0")..SemVer("8.2.0")) {
+                        return false
+                    }
+                }
+            }
+        }
+
+        return true
+    }
+
+    private val versionsMap: Map<String, SemVer> =
+        buildMap {
+            put("Dokka", dgp)
+            put("Kotlin", kgp)
+            put("Gradle", gradle)
+            if (agp != null) put("Android", agp)
+            if (composeGradlePlugin != null) put("Compose", composeGradlePlugin)
+        }
+
+    /**
+     * Join the major and minor versions of each version.
+     * Used by [allTestedVersions] to filter out duplicated patch versions.
+     */
+    private val majorAndMinorVersions: String =
+        versionsMap
+            .mapValues { (_, v) -> "${v.major}.${v.minor}" }
+            .entries
+            .map { (k, v) -> "$k: $v" }
+            .sorted().joinToString(" ")
+
+    /**
+     * Join the versions in a dash-separated-string, for use in file names.
+     */
+    val dashSeparated: String =
+        versionsMap.entries
+            .map { (key, value) -> "$key-$value" }
+            .sorted()
+            .joinToString("_")
+
+    /**
+     * Pretty display name, for use in JUnit test names.
+     */
+    val displayName: String =
+        versionsMap.entries
+            .map { (key, value) -> "$key: $value" }
+            .sorted()
+            .joinToString(", ")
+
+    /**
+     * Compares every version, in order.
+     */
+    override fun compareTo(other: TestedVersions): Int {
+        versionsMap.entries
+            .sortedBy { it.key }
+            .forEach { (key, thisVersion) ->
+                val otherVersion = other.versionsMap[key] ?: return@forEach
+                val cmp = thisVersion.compareTo(otherVersion)
+                if (cmp != 0) {
+                    return cmp
+                }
+            }
+        return 0
+    }
+
+    companion object {
+
+        /** All possible Kotlin Gradle Plugin versions that could be tested. */
+        private val allKgpVersions = listOf(
+            "1.9.0", // required for Compose 1.5.1
+            "1.9.23", // required for Compose 1.5.12
```

Comment:
curious: why do you think we need to test with different compose versions for patch versions of kotlin pre-2.0?

Reply:
The Compose compiler has very strict requirements for specific Kotlin versions, down to the patch level.

https://developer.android.com/jetpack/androidx/releases/compose-kotlin

Without the exact KGP version Compose will throw an error.

Reply:
Yeah, I understand, but can't we just test it for 1.9.25?
I mean, it sounds a bit strange that we test with 3 latest patch versions - one should be enough, no?

Reply:
I'm pretty sure Compose will throw an error if it's used with KGP 1.9.25.

Reply:
Looks like everything works fine with 1.9.24, and in the end, only tests for 1.9.25 (from 1.9.x) are run :)
<img width="865" alt="image" src="https://github.com/user-attachments/assets/e638b14d-848e-424e-897c-71401aa1e21d">

Here, I've added `1.7.1` and `1.8.1` versions to `allComposeGradlePluginVersions` just to check, how it works. And here I see, that `AndroidProjectIT` runs for each compose version, even though it don't uses compose at all.

minor: printing Dokka version there doesn't make sense, as we will always run with single Dokka version (be it current snapshot or override)

I feel like current solution may work now (for current two tests with those specific versions), but it really has like foot-guns and I'm unsure what I think about it.
I like the direction of this PR, really, but it looks too complex for me and so it's hard to understand if it works as intended, and more important, that it will work in the same way when we will add more versions :)

I would like to hear an opinion of someone from KBT on this (e.g second reviewer :)), may be I'm too nitpicking...

Reply:
> And here I see, that `AndroidProjectIT` runs for each compose version, even though it don't uses compose at all.

Thanks, this was a bug, and I've fixed it.


> minor: printing Dokka version there doesn't make sense, as we will always run with single Dokka version (be it current snapshot or override)

My thinking here was to prefer to be explicit. Because the Dokka version can be overridden, I thought it'd be useful to always log the Dokka version. That way, it's easy to see what version of Dokka is being tested.

I looked into adding a filter to only log the version if the Dokka version override was not present, but tbh it was quite complicated and I think more effort than it was worth.